packages/java-parser/src/productions/classes.js

```diff
@@ -662,24 +662,19 @@ function defineRules($, t) {
       return false;
     }
 
-    try {
-      // The {classModifier} is a super grammar of the "interfaceModifier"
-      // So we must parse all the "{classModifier}" before we can distinguish
-      // between the alternatives.
-      $.MANY({
-        GATE: () =>
-          (tokenMatcher($.LA(1).tokenType, t.At) &&
-            tokenMatcher($.LA(2).tokenType, t.Interface)) === false,
-        DEF: () => {
+    let continueLoop = true;
+    while (continueLoop) {
+      if (
+        (tokenMatcher($.LA(1).tokenType, t.At) &&
+          tokenMatcher($.LA(2).tokenType, t.Interface)) === false
+      ) {
+        try {
           $.SUBRULE($.classModifier);
+        } catch (e) {
+          continueLoop = false;
         }
-      });
-    } catch (e) {
-      if (isRecognitionException(e)) {
-        // TODO: add original syntax error?
-        throw "Cannot Identify if the <TypeDeclaration> is a <ClassDeclaration> or an <InterfaceDeclaration>";
       } else {
-        throw e;
+        continueLoop = false;
       }
```

Comment:
It is the same as the OPTION in `isReferenceTypeCastExpression`. If I try to keep the original code, it fails even on:

```java
class T {}
```

with `compilationUnit` as entryPoint in single-sample-runner.js

Reply:
I will try to reproduce the error when running the tests

Reply:
The recording logic gets broken because of an early "return false".
Make sure the `isXXXXX` will not return early during recording phase.
You can use [this flag](https://sap.github.io/chevrotain/documentation/6_5_0/classes/baseparser.html#recording_phase) to assist you.

Basically you want all the DSL methods in the Rules to be invoked during "recording"
If you debug such an "isABC" rule you can see it exits early without getting to the MANY/OPTION and therefor the lookahead function for that rule would not be created
leading to `TypeError: Cannot read property 'call' of undefined`

Reply:
Great ! I refactored the function to avoid early returns